docs/csharp/pattern-matching.md

```diff
@@ -2,31 +2,203 @@
+because of the definitely assigned when true mechanism.
+
+> [!TIP]
+> The samples in this topic use the recommended construct where
+> a pattern match `is` expression definitely assigns the match
+> variable in the `true` branch of the `if` statement.
+> You could reverse the logic by saying `if (!(shape is Square s))`
+> and the variable `s` would be definitely assigned only in the
+> `false` branch. While this is valid C#, it is not recommended
+> because it is more confusing to follow the logic.
+
+These rules mean that you are unlikely to accidentally access the result
+of a pattern match expression when that pattern was not met.
+
+## Using pattern matching `switch` statements
+
+As time goes on, you may need to support other shape types. As the number
+of conditions you are testing grows, you'll find that using the `is` pattern
+matching expressions can become cumbersome. In addition to requiring `if`
+statements on each type you want to check, the `is` expressions are limited
+to testing if the input matches a single type.
+
+As the number of conditions grows, you'll find that the `switch` pattern
+matching expressions becomes a better choice. To get ready to add more
+types of shapes to this library, let's change the existing method to use
+the `switch` pattern match expression:
+
+[!code-csharp[ComputeWithSwitchExpression](../../samples/csharp/PatternMatching/PatternMatching/GeometricUtilities.cs#06_ComputeWithSwitchExpression "Compute with `switch` expression")]
+
+The pattern matching `switch` statement uses familiar syntax to developers
+that have used the traditional C-style `switch` statement. Each `case` is evaluated
+and the code beneath the condition that matches the input variable is
+executed. Code execution cannot "fall through" from one `case` expression
+to the next. Each block following a `case` must end with a `break`, `return`
+or `goto`.
+
+There are important new rules governing the `switch` statement. The restrictions
+on the type of the variable in the `switch` expression have been removed.
+Any type, such as `object` in this example, may be used. The case expressions
+are no longer limited to constant values. Removing that limitation means
+that reordering `switch` sections may change a program's behavior.
+
+When limited to constant values, no more than one `case`
+label could match the value of the `switch` expression. Combine that with the 
+rule that every `switch` section must not fall through to the next section, and 
+it followed that the
+`switch` sections could be rearranged in any order without affecting behavior.
+Now, with more generalized `switch` expressions, the order of each section
+matters. The `switch` expressions are evaluated in textual order. Execution
+transfers to the first `switch` label that matches the `switch` expression.  
+Note that the `default` case will only be executed if no other
+case labels match. The `default` case is evaluated last, regardless
+of its textual order.
```

Comment:
Perhaps also mention what happens if no conditions are met and there is no default?